File: server/src/main/java/com/ultreon/craft/registry/ObjectInit.java
package com.ultreon.craft.registry;

import com.ultreon.craft.registry.DelayedRegister;
import com.ultreon.craft.registry.Registry;
import com.ultreon.craft.registry.RegistrySupplier;

import java.util.function.Supplier;

public abstract class ObjectInit<T> {
    protected final DelayedRegister<T> register;

    public ObjectInit(String namespace, Registry<T> registry) {
        this.register = new DelayedRegister<>(namespace, registry);
    }

    public <C extends T> RegistrySupplier<C> register(String name, Supplier<C> supplier) {
        return this.register.register(name, supplier);
    }
}

File: api-kotlin/src/main/kotlin/com/ultreon/craft/kotlin/registry/ObjectInit.kt
package com.ultreon.craft.kotlin.registry

import com.ultreon.craft.registry.DeferRegistry
import com.ultreon.craft.registry.DeferredElement
import com.ultreon.craft.registry.Registry

abstract class ObjectInit<T>(namespace: String, registry: Registry<T>) {
    private val register: DeferRegistry<T> = DeferRegistry.of(namespace, registry)

    protected fun <C : T & Any> register(name: String?, supplier: () -> C): DeferredElement<C> {
        return register.defer(name!!) {
            supplier()
        }
    }

    fun register() {
        register.register()
    }
}